File: August 2025/21.cpp
class Solution {
  public:
    int maxMinDiff(vector<int>& arr, int k) {
        // Sort the array to simplify difference calculations
        sort(arr.begin(), arr.end());

        // Helper function to check if a minimum difference is feasible
        auto isPossible = [&](int minDiff) {
            int count = 1;
            int lastSelected = arr[0];

            for (int i = 1; i < arr.size(); ++i) {
                if (arr[i] - lastSelected >= minDiff) {
                    count++;
                    lastSelected = arr[i];
                    if (count == k) return true;
                }
            }
            return false;
        };

        // Binary search for the maximum minimum difference
        int low = 0;
        int high = arr.back() - arr.front();
        int result = 0;

        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (isPossible(mid)) {
                result = mid;
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }

        return result;
    }
};

File: August 2025/21.py
class Solution:
    def maxMinDiff(self, arr, k):
        # Sort the array to make difference calculations easier
        arr.sort()

        # Helper function to check if we can select k elements
        # with at least 'min_diff' between any two
        def is_possible(min_diff):
            count = 1
            last_selected = arr[0]

            for i in range(1, len(arr)):
                if arr[i] - last_selected >= min_diff:
                    count += 1
                    last_selected = arr[i]
                    if count == k:
                        return True
            return False

        # Binary search for the maximum minimum difference
        low = 0
        high = arr[-1] - arr[0]
        result = 0

        while low <= high:
            mid = (low + high) // 2
            if is_possible(mid):
                result = mid
                low = mid + 1
            else:
                high = mid - 1

        return result
